### PR TITLE
Fixing issues with imbalanced datasets

### PR DESCRIPTION
**Pull request by ArlindKadra**
This PR fixes issue #195

### CODE REVIEW

**Review by ravinkohli — CHANGES_REQUESTED** [review]
Hey Arlind, thanks for the PR. Its essential in the corner case when we have all nans in train but not in test and vice versa. But, could you add tests for the validator to ensure we can handle the case properly. It would be great if we could test the two cases- 1. where we have all NaNs in train and not in test and 2. where we have all NaNs in test but not in train

**Comment by ArlindKadra** [comment]
> Hey Arlind, thanks for the PR. Its essential in the corner case when we have all nans in train but not in test and vice versa. But, could you add tests for the validator to ensure we can handle the case properly. It would be great if we could test the two cases- 1. where we have all NaNs in train and not in test and 2. where we have all NaNs in test but not in train

Yep, I will do just that today. But overall, I wanted to get a quick idea on how the problem is handled now, in case I missed something on the fix.

File: autoPyTorch/data/tabular_feature_validator.py [hunk 1]
```diff
@@ -146,32 +151,35 @@ def transform(
...
+            for column in self.null_columns:
+                # The column is not null, make it null since it was null in fit.
+                if not X[column].isna().all():
+                    X[column] = np.NaN
+                X[column] = pd.to_numeric(X[column])
+
+            # for the test set, if we have columns with only null values
+            # they will probably have a numeric type. If these columns were not
+            # with only null values in the train set, they should be converted
+            # to the category type.
```
Comment (ravinkohli) [inline autoPyTorch/data/tabular_feature_validator.py:167]:
```suggestion
            # to the type it had while fitting.
```

I don't think we should wait for the tests to merge in case you choose to incorporate this suggestion

File: autoPyTorch/data/tabular_feature_validator.py [hunk 2]
```diff
@@ -414,7 +422,7 @@ def infer_objects(self, X: pd.DataFrame) -> pd.DataFrame:
                     # In the case train data was interpreted as int
                     # and test data was interpreted as float, because of 0.0
                     # for example, honor training data
-                    X[key] = X[key].applymap(np.int64)
+                    X[key] = X[key].astype(np.int64)
```
Comment (ravinkohli) [inline autoPyTorch/data/tabular_feature_validator.py:425]: should we remove the if-else statement? as we are treating them equally

Comment (ArlindKadra) [reply autoPyTorch/data/tabular_feature_validator.py]: Yep, nice catch.

File: autoPyTorch/data/tabular_feature_validator.py (hunk 1 above)
Comment (ArlindKadra) [reply autoPyTorch/data/tabular_feature_validator.py]: I also changed the comment accordingly.

File: test/test_data/test_feature_validator.py [hunk 3]
```diff
@@ -515,11 +518,73 @@ def test_featurevalidator_new_data_after_fit(openml_id,
...
+    assert train_feature_types == ['categorical', 'numerical', 'numerical']
+
+    transformed_X_test = validator.transform(X_test)
+    transformed_X_test = pd.DataFrame(transformed_X_test)
+    null_columns = []
+    for column in transformed_X_test.columns:
+        if transformed_X_test[column].isna().all():
+            null_columns.append(column)
+
+    assert null_columns == [1]
```
Comment (ravinkohli) [inline test/test_data/test_feature_validator.py:590]: why is null_columns not empty here?

Comment (ArlindKadra) [reply test/test_data/test_feature_validator.py:590]: Because for the test set, column 'A' will have only null values, however, it will have a numeric type. While column 'C' will have 1, as it will detect correctly that it has 2 categories from the train set.

Comment (ravinkohli) [reply test/test_data/test_feature_validator.py:590]: I see, sorry I got confused as I expected none to be Null, I don't know why

**Review by ravinkohli — APPROVED** [review]
